gossip/evmstore/store.go

```diff
@@ -126,6 +126,31 @@ func (s *Store) RebuildEvmSnapshot(root common.Hash) {
 	s.Snaps.Rebuild(root)
 }
 
+// Commit changes.
+func (s *Store) CleanCommit(block iblockproc.BlockState) error {
+	// Don't need to reference the current state root
+	// due to it already be referenced on `Commit()` function
+	triedb := s.EvmState.TrieDB()
+	stateRoot := common.Hash(block.FinalizedStateRoot)
+	current := uint64(block.LastBlock.Idx)
+	// Garbage collect all below the current block
+	for !s.triegc.Empty() {
+		root, number := s.triegc.Pop()
```

Comment:
If we dereference all roots but the latest one, then it conflicts a little with `evmstore.Store.Flush`. `evmstore.Store.Flush` tries to commit `head`, `head-1`, `head-31` roots. All of them but `head` will already be dereferenced, so we can simplify it and commit only the `head` inside `Flush`. `Flush` also won't need the `getBlock` argument anymore

Reply:
Also, if we dereference all roots bit the latest one, it will make them inaccessible by API calls (if I understand correctly). Then, we don't respect the `TriesInMemory` constant. As far as I understand, we should either make `TriesInMemory`=1 or not dereference roots inside `CleanCommit`

Reply:
@uprendis Yes, I think in our case, we only need to commit only the `head` inside `Flush` due to go-opera don't need to re-org if I understand correctly. We might change `TriesInMemory`=1. I will consolidate all the related things and commit the code afterward. Thanks!

Reply:
@uprendis by the way, on the 2nd thought, if we periodic call the commit every 10 seconds with `period: 10 * time.Millisecond` in the `PeriodicFlusher` then I think we don't have any effective way to pruning the state on mem-cache. Or I mis-understand smth on here.

Reply:
We don't prune every 10 ms, we check if a commit is needed every 10ms. flushable.Store tracks the size of in-memory diff. We sum those sizes for every DB and compare it to a configured threshold. Practically we commit every ~30-100 blocks on mainnet.
But yes, it does mean that geth's pruning strategy may be not very effective for us, because it's based only on avoiding writing MPT data (unless node stops, or `TrieTimeLimit` passed since prev flush, or `Cap` is called and dirties map is too large). All we can archive with geth's approach is that we will flush every ~30-100th root rather than every root (which is still better than nothing, but not ideal in terms of data pruning). To make a more greedy pruning, we would need to not only avoid writing MPT data, but also erase previously written MPT data if it's not referenced anymore by latest root(s)

Reply:
Yes, but it seems that the `s.dbs.NotFlushedSizeEst()` quickly return a big number (seems accumulating) then the commit happen very frequency via `mayCommit()` method call

Reply:
@uprendis I debug log and found that with current **snapsync**, the latest block state is always at block number 1

```
// GetBlockState retrieves the latest block state
func (s *Store) GetBlockState() iblockproc.BlockState {
	return *s.getBlockEpochState().BlockState
}
```

So the below **commitEVM** invoke has no meaning for state commit for the current code on **develop** branch

```
	// TODO: prune old MPTs in beginnings of committed sections
	if !s.store.cfg.EVM.Cache.TrieDirtyDisabled {
		s.store.commitEVM(true)
	}
```

**_And I think we also cannot make the trie state pruning by using the current triegc due to in that way, we need to record every block's state root into the triegc queue then doing the pruning later._** but with snapsync, we didn't record the state root into the triegc.

Reply:
```
I debug log and found that with current snapsync, the latest block state is always at block number 1
```
Yeah, that's right. While snapsync hasn't finished, current block number won't move. After snapsync, we'll jump from an epoch which we had before the snapsync, to the new epoch
Yes, Service.commit during snapsync will only flush the content of flushable.Store's. s.store.commitEVM will practically do nothing

Reply:
```
And further more I think we also cannot make the trie state pruning by using the current triegc due to in that way, we need to record every block's state root into the triegc queue then doing the pruning later. but with snapsync, we didn't record the state root into the triegc.
```
Hmm, I don't quite understand. After snapsync, the MPTs will already be pruned, because we downloaded only latest roots. After snapsync, triegc should work (not effectively as it'll commit every ~30-100th root)?

Reply:
> 
> Hmm, I don't quite understand. After snapsync, the MPTs will already be pruned, because we downloaded only latest roots. After snapsync, triegc should work (not effectively as it'll commit every ~30-100th root)?

Yes, I were thinking about when the snapsync in-progress so I though like that. After snapsync it be work like you said.